### ISSUE TITLE
Instance segmentation

### ISSUE THREAD

**Issue by ai8049520** [tpnam0901/QTSeg#4]
Hello! Thank you for your great model!

I train my dataset with 1 small pathology and got promising results. My question is how can I prepare it for instance segmentation? Can I save my masks with instance "color" and retrain to get instances in prediction?

**Comment by tpnam0901** [comment]
Hello @ai8049520, thank you for your interest in our project. 

Currently, it's hard to make the current repository support instance segmentation. But in my experience, here is what you can work around.

**First - add an object detection head to the model** 
To train instance segmentation, you need to train another prediction head for the bounding box. This bbox head can be attached after the ```current_feature``` in [mask_decoder](https://github.com/tpnam0901/QTSeg/blob/cde55eb04c5b745158eb514b15bbe4dd01fd08c8/src/networks/decoder/models.py#L168). We have three outputs of current_feature, so we need three bbox heads.

**Secondly - build  loss function  and anchor bounding box for dataloader**
You need to define the target for these boxes to calculate mean square error or mean absolute errors. You can refer to this [repo](https://github.com/yhenon/pytorch-retinanet). The performance of your model depends on how you process the anchors. 

**Thirdly - change the inference process**
You may need to change the inference by applying non-max suppression after the output of models to remove the overlapping box. Then, you can map the prediction box with its mask. 

**However, this will be a lot of work, and I am not sure we can develop from this repo.** If your input image does not have an overlapping object, you can just use image processing to find the connected mask after having the prediction for finding the object.

**P/s:** ```Can I save my masks with instance "color" and retrain to get instances in prediction?``` Our model expects the target mask in the format of [0,1,2,...,num_classes]. If you save the model with two values [0,255], it will be converted back to [0,1]. We expect that class ```0 is background``` and 1,2,... num_classes are the target labels. For example, 1-cat, 2-dog, etc. If you save your mask with labels ```0-background, 1-cat, 2-cat,...num_cats``` and change the ```num_masks``` to ```num_cats``` I'm not sure it will work because it's not the proper way to do this :). Also, you need to consider the number of cats in the image.

**Comment by ai8049520** [comment]
Thank you for your detailed answer! I use "1" for pathology class in binary mask , and aftre prdect use postprocessing for instance detection to compare with other models (yolo11-seg) and metrics. I see that addition of detector heads and loses is quite complicated task and needs a lot of efforts. Thank you again for this model!

By the way, why some metrics like Precision and Recall are commented out in code?  
 def compute(self):
        return {
            "mae": np.mean(self.MAE) * 100,
            # "recall": np.mean(self.Recall) * 100,
            # "precision": np.mean(self.Precision) * 100,
            "acc": np.mean(self.Accuracy) * 100,
            "dice": np.mean(self.Dice) * 100,
            "iou": np.mean(self.IoU_polyp) * 100,
        }

**Comment by tpnam0901** [comment]
@ai8049520 You can also uncomment these metrics because I do not use them to compare with previous works. I just commented out these codes for clean logging information. In addition, with my training script, they will also save the best value for "recall" and "precision," which is useless and costs our memory.
